### PR TITLE
Add Command to add Session 

### PR DESCRIPTION
**Pull request by chuayijing**

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddSessionCommand.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            + PREFIX_DATE + "19-02-2020 "
+            + PREFIX_MOD_CODE + "CS2103T";
+
+    public static final String MESSAGE_SUCCESS = "New session added: %1$s";
+    public static final String MESSAGE_DUPLICATE_SESSION = "This session already exists in the address book";
+
+    private final Session toAdd;
+
+    /**
+     * Creates an AddCommand to add the specified {@code Session}
```
Comment (PotatoCombat) [inline src/main/java/seedu/address/logic/commands/AddSessionCommand.java]: Creates an AddSessionCommand!

File: src/main/java/seedu/address/logic/parser/CliSyntax.java [hunk 2]
```diff
@@ -11,5 +11,11 @@
     public static final Prefix PREFIX_EMAIL = new Prefix("e/");
     public static final Prefix PREFIX_MATRIC = new Prefix("m/");
     public static final Prefix PREFIX_TAG = new Prefix("t/");
-
+    public static final Prefix PREFIX_STARTTIME = new Prefix("s/");
+    public static final Prefix PREFIX_ENDTIME = new Prefix("e/");
+    public static final Prefix PREFIX_DATE = new Prefix("d/");
+    public static final Prefix PREFIX_RECUR = new Prefix("[-r]");
+    public static final Prefix PREFIX_MOD_CODE = new Prefix("m/");
```
Comment (PotatoCombat) [inline src/main/java/seedu/address/logic/parser/CliSyntax.java:18]: Just leaving a note here. PREFIX_MOD_CODE might already have been defined when the Module class was created.

File: src/main/java/seedu/address/model/Model.java [hunk 3]
```diff
@@ -52,6 +57,26 @@
...
     ReadOnlyAddressBook getAddressBook();
 
+    /**
+     * Returns true if a session with the same identity as {@code session} exists in the address book.
+     */
+    boolean hasSession(Session session);
+
+    /**
+     * Adds the given person.
+     * {@code person} must not already exist in the address book.
```
Comment (PotatoCombat) [inline src/main/java/seedu/address/model/Model.java:66]: @code session!

File: src/main/java/seedu/address/model/session/Session.java [hunk 4]
```diff
@@ -50,6 +50,16 @@ public Session(LocalTime startTime, LocalTime endTime, LocalDate date, SessionTy
         this.description = description;
     }
 
+    public Session() {}
```
Comment (PotatoCombat) [inline src/main/java/seedu/address/model/session/Session.java]: Should Session have an empty constructor? 😳

File: src/main/java/seedu/address/model/session/Session.java [hunk 5]
```diff
@@ -50,6 +50,16 @@ public Session(LocalTime startTime, LocalTime endTime, LocalDate date, SessionTy
...
+
+    /**
+     * Returns true if both session of the same name have at least one other identity field that is the same.
+     * This defines a weaker notion of equality between two session.
+     */
+    public boolean isSameSession(Session s) {
+        return false;
+    }
```
Comment (PotatoCombat) [inline src/main/java/seedu/address/model/session/Session.java]: Actually, if `Session` has already been defined, this shouldn't be a stub.

File: src/main/java/seedu/address/model/session/Session.java [hunk 6]
```diff
@@ -91,4 +101,12 @@ public String getDescription() {
...
         return Duration.between(this.startTime, this.endTime);
     }
+
+    /**
+     * Returns true if both sessions have the same identity and data fields.
+     */
+    @Override
+    public boolean equals(Object other) {
+        return other == this;
+    }
```
Comment (PotatoCombat) [inline src/main/java/seedu/address/model/session/Session.java:109]: Actually, if Session has already been defined, this shouldn't be a stub.

File: src/main/java/seedu/address/model/session/UniqueSessionList.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+ * @see Session#isSameSession(Session)
+ */
+public class UniqueSessionList implements Iterable<Session> {
+
+    private final ObservableList<Session> internalList = FXCollections.observableArrayList();
+    private final ObservableList<Session> internalUnmodifiableList =
+            FXCollections.unmodifiableObservableList(internalList);
+
+    /**
+     * Returns true if the list contains an equivalent student as the given argument.
```
Comment (PotatoCombat) [inline src/main/java/seedu/address/model/session/UniqueSessionList.java]: ... an equivalent session!

File: src/main/java/seedu/address/model/session/UniqueSessionList.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+     */
+    public boolean contains(Session toCheck) {
+        requireNonNull(toCheck);
+        return internalList.stream().anyMatch(toCheck::isSameSession);
+    }
+
+    /**
+     * Adds a student to the list.
+     * The student must not already exist in the list.
```
Comment (PotatoCombat) [inline src/main/java/seedu/address/model/session/UniqueSessionList.java]: Session not student!

File: src/main/java/seedu/address/model/session/UniqueSessionList.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+        requireAllNonNull(sessions);
+        if (!sessionsAreUnique(sessions)) {
+            throw new DuplicateSessionException();
+        }
+
+        internalList.setAll(sessions);
+    }
+
+    /**
+     * Returns the backing list as an unmodifiable {@code ObservableList}.
```
Comment (PotatoCombat) [inline src/main/java/seedu/address/model/session/UniqueSessionList.java]: Hmm... maybe can change this to "returns the session list"

File: src/test/java/seedu/address/logic/commands/AddCommandTest.java [hunk 10]
```diff
@@ -147,6 +148,44 @@ public void setPerson(Student target, Student editedStudent) {
         public void updateFilteredPersonList(Predicate<Student> predicate) {
             throw new AssertionError("This method should not be called.");
         }
+
+        @Override
+        public ObservableList<Session> getFilteredSessionList() {
+            throw new AssertionError("This method should not be called.");
+        }
+
+        @Override
```
Comment (PotatoCombat) [inline src/test/java/seedu/address/logic/commands/AddCommandTest.java]: Actually these methods should be in a separate test class called AddSessionCommandTest. You can refer to the format of the original AddCommandTest!

File: build.gradle [hunk 11]
```diff
@@ -64,6 +64,8 @@ dependencies {
     testImplementation group: 'org.junit.jupiter', name: 'junit-jupiter-api', version: jUnitVersion
 
     testRuntimeOnly group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: jUnitVersion
+
+    testCompileOnly('org.apiguardian:apiguardian-api:1.0.0')
```
Comment (PotatoCombat) [inline build.gradle]: Double check if this dependency is still required?

**Review by PotatoCombat — COMMENTED** [review]
The stub methods in `Session` can actually be implement at this point!

Also I was able to build without the api guardian dependency. Maybe there was some other reason for the warning you saw?

Last of all, just need to double check some of your javadocs.

File: src/main/java/seedu/address/model/Model.java [hunk 12]
```diff
@@ -53,9 +54,9 @@
     ReadOnlyAddressBook getAddressBook();
 
     /**
-     * Returns true if a person with the same identity as {@code person} exists in the address book.
+     * Returns true if a session with the same identity as {@code session} exists in the address book.
      */
-    boolean hasPerson(Person person);
+    boolean hasPerson(Session session);
```
Comment (Eclmist) [inline src/main/java/seedu/address/model/Model.java]: Should this be `hasSession` instead?

File: src/main/java/seedu/address/model/Model.java [hunk 13]
```diff
@@ -6,13 +6,14 @@
...
 import seedu.address.model.person.Person;
+import seedu.address.model.session.Session;
 
 /**
  * The API of the Model component.
  */
 public interface Model {
     /** {@code Predicate} that always evaluate to true */
-    Predicate<Person> PREDICATE_SHOW_ALL_PERSONS = unused -> true;
+    Predicate<Person> PREDICATE_SHOW_ALL_SESSION = unused -> true;
```
Comment (Eclmist) [inline src/main/java/seedu/address/model/Model.java]: `Model` seems to be used by the student list as well, should `person` be overridden to `session` or should a new one be added for `session`?

**Comment by PotatoCombat** [comment]
Remember to fixup any check style related errors! If your builds fail, you should fixup the commit, then update your feature branch with a force push. If it helps, try to fix style errors as you code too (it helps to run checkstyle before making a commit, and it's faster than waiting for the builds).

The idea is to have a clear commit history. It can be confusing to follow the history if there are too many minor commits such as checkstyle fixes.

Just did a bit of fixup to your branch. You can find these changes on fix/session-add. If you want you can make a backup of this branch, then run:
```
# Replaces your feature branch with the fix branch
git fetch origin
git checkout feature/session-add
git reset --hard origin/fix/session-add
```
You'll need to change the commit author back to you. Here's a script to help you:

- Download [resetAuthor.txt](https://github.com/AY1920S2-CS2103T-W17-4/main/files/4335393/resetAuthor.txt) to your local tP directory.

- Open a terminal, and cd to your the local tP directory.
- Then execute `bash resetAuthor.txt`, following the other instructions.

Remember to delete this file when you're done! It's not meant to be committed!

**Review by PotatoCombat — COMMENTED** [review]
Just need the stubs to be implemented!

File: src/main/java/seedu/address/model/Model.java (hunk 13 above)
Comment (PotatoCombat) [reply src/main/java/seedu/address/model/Model.java]: The new changes keep both now!

File: src/main/java/seedu/address/model/AddressBook.java [hunk 14]
```diff
@@ -83,6 +83,10 @@ public boolean hasPerson(Person person) {
         return persons.contains(person);
     }
 
+    /**
+     * Adds a person to the address book.
+     * The person must not already exist in the address book.
+     */
```
Comment (Eclmist) [inline src/main/java/seedu/address/model/AddressBook.java]: This commit content doesn't match commit message of "Add UniqueSessionList in Address Book"

Comment (Eclmist) [inline src/main/java/seedu/address/model/AddressBook.java]: Commit message doesn't match the commit content

**Comment by PotatoCombat** [comment]
The pull request got broken again for some reason?
You'll need to create a new PR!

**Pull request closed without merging**